debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java

```diff
@@ -177,27 +176,25 @@ protected void storeRecord(HistoryRecord record) throws DatabaseHistoryException
     }
 
     @Override
-    protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<HistoryRecord> records) {
+    protected void recoverRecords(Consumer<HistoryRecord> records) {
         try (KafkaConsumer<String, String> historyConsumer = new KafkaConsumer<>(consumerConfig.asProperties());) {
             // Subscribe to the only partition for this topic, and seek to the beginning of that partition ...
             logger.debug("Subscribing to database history topic '{}'", topicName);
             historyConsumer.subscribe(Collect.arrayListOf(topicName));
 
-            // Explicitly seek to the beginning of all assigned topic partitions ...
-            logger.debug("Seeking Kafka consumer to beginning of all assigned topic partitions");
-            historyConsumer.seekToBeginning(Collections.emptyList());
-
             // Read all messages in the topic ...
-            int remainingEmptyPollResults = this.recoveryAttempts;
-            Map<Long, Long> offsetsByPartition = new HashMap<>();
-            while (remainingEmptyPollResults > 0) {
+            long lastProcessedOffset = -1;
+            long endOffset = getEndOffsetOfDbHistoryTopic(historyConsumer);
+            logger.debug("End offset of database history topic is {}", endOffset);
+
+            // read the topic until the end
+            while (lastProcessedOffset < endOffset - 1) {
```

Comment:
Do we want to still limit the number of retries here?

I was thinking leaving the `remainingEmptyPollResults` functionality, but throw a fatal Exception if we exit the loop without reaching the endOffset, leaving the connector in a Failed state, rather than just trying to recover indefinitely. This escalates the problem to the connector status that something went wrong.

Maybe increase the default for `RECOVERY_POLL_ATTEMPTS` in conjunction with the new recovery algorithm.

Reply:
Yeah, I was wondering that, too.

OTOH I'm not sure how actionable such failure would be. Eventually, we'll _need_ to read up the end, so would the user have any other way than increasing that number of retries? Also what'd be a good default value for it? But I also see how the - potentially - unbounded - loop can be discomforting.

Reply:
Regarding limiting the number of times we go through this loop: the only time we'll run into problems is if there is a concurrent producer updating the topic while we're trying to consume it. That's a pathological case that might only be true if multiple connectors are using the same topic. In fact, we could detect that situation and fail with a useful error message. Of course, we wouldn't be able to always detect this (e.g., the other connector might not be recording any DDL statements at the time), but I still think it might be valuable.

In fact, I think the logic in this PR currently doesn't really handle the case where the end offset might be changing. To detect that, we should call `getEndOffsetOfDbHistoryTopic(historyConsumer)` within the loop and compare it to `endOffset`, reporting an error if the new end offset is larger than the previous end offset. However, after we do this, we may want to update `endOffset` just in case it is *smaller* because some messages might have been lost in the broker. (I'm not really sure whether that's even possible, but it's trivial to do.)

Now, what else might go wrong within the loop such that we can't get to the end offset (even if it is fixed)? The `poll` method is the big culprit here, and it might return 0 records because it times out and hasn't received any records when indeed we know there are more records. So we could we limit this logic, but do we want to do it by attempts or by time? Either way, we're waiting for the consumer to return records that should be there, so we could continue for a relatively long time.

Reply:
In the (unlikely) indefinite loop case, is there any way the User would know it's stuck here in order to take appropriate action? 

I'm not sure what the best default would be for RECOVERY_POLL_ATTEMPTS. I believe we set ours to 20, but that was to try to workaround this exact reported issue.

Reply:
@rhauch I agree that if we go with retries it would be reasonable to have this continue to retry for a very long time before failing. 

If we have not reached the end of the topic but are getting no records from `poll()` for many minutes, we know *something* is definitely wrong, so I'm mostly interested in getting that message to the user / admin in some way so they can know where in the startup/recovery process it's getting stuck in order to take action.

I believe even a user could simply set the RECOVERY_POLL_INTERVAL_MS too small and could run into this inadvertently.

Reply:
How about not _limiting_ the retries but raising a WARN message if we can't reach the end after a while / after a few tries? That way users will have a sense that something is wrong. Of course not as apparent as a connector in FAILED state.

Regarding moving the retrieval of the end offsets _into_ the loop, I had that originally. But it seems a costly operation which increased execution time for me (e.g. KafkaDatabaseHistoryTest went up by 10 sec). Is it worth that only to discover that pathological case of a moving end offset? Normally, the end offset should be fixed, meaning we'd re-read the same value in the loop repeatedly.

Reply:
@gunnarmorling The WARN message would be fine as well.

Reply:
I've reworked that last commit. It'll raise an exception now if it reaches the max retry count before reaching the end of the topic, so the connector will go to FAILED then. I've increased the default to 100, too. The end offset retrieval is part of the loop now, so we'll detect changes by other connectors in some cases at least.